xds/src/main/java/io/grpc/xds/XdsComms2.java

```diff
@@ -0,0 +1,247 @@
+                    clusterLoadAssignment =
+                        value.getResources(0).unpack(ClusterLoadAssignment.class);
+                  } catch (InvalidProtocolBufferException | RuntimeException e) {
+                    cancelRpc("Received invalid EDS response", e);
+                    adsStreamCallback2.onError();
+                    scheduleRetry();
+                    return;
+                  }
+
+                  helper.getChannelLogger().log(
+                      ChannelLogLevel.DEBUG,
+                      "Received an EDS response: {0}", clusterLoadAssignment);
+                  adsStreamCallback2.onEdsResponse(clusterLoadAssignment);
+                }
+              }
+            }
+
+            helper.getSynchronizationContext().execute(new HandleResponseRunnable());
+          }
+
+          @Override
+          public void onError(Throwable t) {
+            helper.getSynchronizationContext().execute(
+                new Runnable() {
+                  @Override
+                  public void run() {
+                    closed = true;
+                    if (cancelled) {
+                      return;
+                    }
+                    adsStreamCallback2.onError();
+                    scheduleRetry();
+                  }
+                });
+          }
+
+          @Override
+          public void onCompleted() {
+            onError(Status.INTERNAL.withDescription("Server closed the ADS streaming RPC")
+                .asException());
+          }
+
+          // run in SynchronizationContext
+          void scheduleRetry() {
+            if (channel.isShutdown()) {
+              return;
+            }
+
+            checkState(
+                cancelled || closed,
+                "Scheduling retry while the stream is neither cancelled nor closed");
+
+            checkState(
+                adsRpcRetryTimer == null, "Scheduling retry while a retry is already pending");
+
+            class AdsRpcRetryTask implements Runnable {
+              @Override
+              public void run() {
+                adsRpcRetryTimer = null;
+                refreshAdsStream();
+              }
+            }
+
+            if (firstEdsResponseReceived) {
+              // Reset the backoff sequence if balancer has sent the initial response
+              adsRpcRetryPolicy = backoffPolicyProvider.get();
+              // Retry immediately
+              helper.getSynchronizationContext().execute(new AdsRpcRetryTask());
+              return;
+            }
+
+            adsRpcRetryTimer = helper.getSynchronizationContext().schedule(
+                new AdsRpcRetryTask(),
+                adsRpcRetryPolicy.nextBackoffNanos() - retryStopwatch.elapsed(TimeUnit.NANOSECONDS),
+                TimeUnit.NANOSECONDS,
+                helper.getScheduledExecutorService());
+          }
+        };
+
+    boolean cancelled;
+    boolean closed;
+
+    AdsStream(AdsStreamCallback2 adsStreamCallback2) {
```

Comment:
It's better to have `start/stop` methods instead of starting right in constructor.